lib/angstrom.ml

```diff
@@ -455,10 +455,11 @@ let choice ?(failure_msg="no more choices") ps =
   List.fold_right (<|>) ps (fail failure_msg)
 
 let fix_direct f =
-  let rec p = lazy (f r)
+  let rec p = ref None
   and r = { run = fun buf pos more fail succ ->
-    (Lazy.force p).run buf pos more fail succ }
+    (Option.get !p).run buf pos more fail succ }
   in
+  p := Some (f r);
   r
```

Comment:
From a distance this looks like it could be made (slightly) more efficient with

```
let p = ref { run = fun buf pos more fail succ -> failwith "Angstrom.fix_direct" }
let r = { run = fun buf pos more fail succ -> !p.run buf pos more fail succ }
p := f r;
!p
```

which saves the `Option.get` on internal calls.

Reply:
Thanks, I've applied this suggestion, and took a bit further by preallocating the record:
```
+let notset = { run = fun _buf _pos _more _fail _succ -> failwith "Angstrom.fix_direct not set" }
```

Reply:
I’d prefer this to maintain compatibility with older stdlibs (ci is failing because of this), and to save the indirection.

Reply:
Nitpick: the commit message says that `notset` is never invoked, but it may be invoked if `f` calls its argument in an non-delayed way. For example, `fix (fun p -> parse_string Prefix p ""; p)` would probably call `notset`. This is okay, it is a failure case that existed before and would raise Lazy.Undefined; but it means that users can (in theory) observe the failure, which should be handled like a programming error by the user and not an this-can-never-possibly-happen situation.

Reply:
I am ok merging the code as-is but will wait for people to be happy with commit history and/or docs before I do so. Just ping the PR when I should look.

Reply:
Thanks, I've pushed an updated commit message, @gasche does it look good?

About the CI I'd suggest merging https://github.com/inhabitedtype/angstrom/pull/222 or some variant of it (lets continue the discussion there), as it will fix most failures in the ocaml-ci. Here is a branch I pushed to my fork with this PR + PR222 to test on ocaml-ci: https://ocaml.ci.dev/github/edwintorok/angstrom/commit/39616024841e1ba98d83b641cdeee288bd39ee9f